diff --git a/src/Surface.zig b/src/Surface.zig
index 2c4edc497..bbd476d4a 100644
--- a/src/Surface.zig
+++ b/src/Surface.zig
@@ -5655,7 +5655,7 @@ fn writeScreenFile(
 ) !void {
     // Create a temporary directory to store our scrollback.
     var tmp_dir = try internal_os.TempDir.init();
-    errdefer tmp_dir.deinit();
+    defer tmp_dir.deinit();
 
     var filename_buf: [std.fs.max_path_bytes]u8 = undefined;
     const filename = try std.fmt.bufPrint(
@@ -5724,7 +5724,6 @@ fn writeScreenFile(
 
         const sel = sel_ orelse {
             // If we have no selection we have no data so we do nothing.
-            tmp_dir.deinit();
             return;
         };
 
diff --git a/src/os/TempDir.zig b/src/os/TempDir.zig
index c3d344d9d..b09396223 100644
--- a/src/os/TempDir.zig
+++ b/src/os/TempDir.zig
@@ -61,5 +61,6 @@ pub fn name(self: *TempDir) []const u8 {
 /// directory.
 pub fn deinit(self: *TempDir) void {
+    defer self.parent.close();
     self.dir.close();
     self.parent.deleteTree(self.name()) catch |err|
         log.err("error deleting temp dir err={}", .{err});
@@ -68,4 +68,9 @@ test {
     const testing = std.testing;
 
+    const tmp_path = try file.allocTmpDir(std.heap.page_allocator);
+    defer file.freeTmpDir(std.heap.page_allocator, tmp_path);
+    var root = try std.fs.cwd().openDir(tmp_path, .{});
+    defer root.close();
+
     var td = try init();
     errdefer td.deinit();
@@ -81,4 +86,4 @@ test {
     // Should be deleted after we deinit
     td.deinit();
-    try testing.expectError(error.FileNotFound, td.parent.openDir(nameval, .{}));
+    try testing.expectError(error.FileNotFound, root.openDir(nameval, .{}));
 }
 
